File: nginx_test/was-4-springboot/src/main/java/com/nginx/test/service/MessageService.java
package com.nginx.test.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

@Slf4j
@Service
@RequiredArgsConstructor
public class MessageService {

    private final KafkaTemplate<String, String> kafkaTemplate;

    public String send(String topic, String key, String message) {
        String messageId = UUID.randomUUID().toString();
        String messageKey = key != null ? key : messageId;

        String wrappedMessage = String.format(
                "{\"id\":\"%s\",\"source\":\"was-4-springboot\",\"timestamp\":\"%s\",\"payload\":%s}",
                messageId, Instant.now(), message
        );

        CompletableFuture<SendResult<String, String>> future =
                kafkaTemplate.send(topic, messageKey, wrappedMessage);

        future.whenComplete((result, ex) -> {
            if (ex != null) {
                log.error("Failed to send message: id={}, topic={}, error={}",
                        messageId, topic, ex.getMessage());
            } else {
                log.info("Message sent: id={}, topic={}, partition={}, offset={}",
                        messageId, topic,
                        result.getRecordMetadata().partition(),
                        result.getRecordMetadata().offset());
            }
        });

        return messageId;
    }

    public Map<String, Object> healthCheck() {
        Map<String, Object> status = new HashMap<>();
        try {
            var metrics = kafkaTemplate.metrics();
            status.put("status", "UP");
            status.put("metrics_count", metrics.size());
        } catch (Exception e) {
            status.put("status", "DOWN");
            status.put("error", e.getMessage() != null ? e.getMessage() : "Unknown error");
        }
        return status;
    }
}

File: nginx_test/was-3-springboot/src/main/kotlin/com/nginx/test/service/MessageService.kt
package com.nginx.test.service

import io.micrometer.observation.annotation.Observed
import org.slf4j.LoggerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.support.SendResult
import org.springframework.stereotype.Service
import java.util.UUID
import java.util.concurrent.CompletableFuture

@Service
@Observed(name = "message.service")
class MessageService(
    private val kafkaTemplate: KafkaTemplate<String, String>
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    fun send(topic: String, key: String?, message: String): String {
        val messageId = UUID.randomUUID().toString()
        val messageKey = key ?: messageId

        val wrappedMessage = """
            {
                "id": "$messageId",
                "source": "was-3-springboot",
                "timestamp": "${java.time.Instant.now()}",
                "payload": $message
            }
        """.trimIndent()

        val future: CompletableFuture<SendResult<String, String>> =
            kafkaTemplate.send(topic, messageKey, wrappedMessage)

        future.whenComplete { result, ex ->
            if (ex != null) {
                logger.error("Failed to send message: id=$messageId, topic=$topic, error=${ex.message}")
            } else {
                logger.info(
                    "Message sent: id=$messageId, topic=$topic, " +
                    "partition=${result.recordMetadata.partition()}, " +
                    "offset=${result.recordMetadata.offset()}"
                )
            }
        }

        return messageId
    }

    fun healthCheck(): Map<String, Any> {
        return try {
            // Simple health check by getting cluster info
            val metrics = kafkaTemplate.metrics()
            mapOf(
                "status" to "UP",
                "metrics_count" to metrics.size
            )
        } catch (e: Exception) {
            mapOf(
                "status" to "DOWN",
                "error" to (e.message ?: "Unknown error")
            )
        }
    }
}
